### PR TITLE
feat(hierarchy-list): add controlled expandedIds prop

### LINKED ISSUE #3037: [HierarchyList] Expand/Collapse all or some rows
### What package is this for?

- [x] React
- [ ] Angular

### Summary

I want to be able to expand/collapse all levels of a HierarchyList (basically pass an array of row Ids as a prop to allow expanding or collapsing of rows).
Using `defaultExpandedIds` only expands specified rows on the first render. I can change the value of that prop by changing a state, but it doesn’t trigger a change in the collapse/expanded state of the rows.

When clicking the Open All button
Screen Shot 2021-11-04 at 9 17 13 AM

All rows specified in the prop are expanded
Screen Shot 2021-11-04 at 9 17 43 AM

### Specific timeline issues / requests

The sooner the better. We have a product release coming up in January, so would be great to have the feature in before then.

### Available extra resources

I am available to test the feature

### PR DESCRIPTION
**Pull request by kevinsperrine**
Closes #3037 

**Summary**

- Adds the ability to explicitly control which ids are expanded or collapsed in the HierarchyList. 
- Also adds an `onExpandedChange` callback that passes the array to the user, so they can know what ids are currently expanded

**Change List (commits, features, bugs, etc)**

- adds `expandedIds` prop to control which ids in the list are expanded or collapsed
- adds `onExpandedChange` callback to let the consumer know the current internal state of the expansions.
- added tests to confirm

**Acceptance Test (how to verify the PR)**

- Check https://deploy-preview-3081--carbon-addons-iot-react.netlify.app/?path=/story/1-watson-iot-hierarchylist--with-overflow-menu
- confirm Chicago White Sox is expanded
- Add another team to the expandedIds array knob (comma separated no space--ie. Chicago White Sox,New York Mets)
- Check https://deploy-preview-3081--carbon-addons-iot-react.netlify.app/?path=/story/1-watson-iot-hierarchylist--with-default-expanded-ids
- confirm the white sox and yankees are expanded by default
- Add a different team in the expandedIds array knob and confirm that team is expanded and the white sox and yankees close

**Regression Test (how to make sure this PR doesn't break old functionality)**

- check for errors with defaultExpandedIds story
- check expanding and collapsing on click still works.
- expanded state should only change when the `expandedIds` prop is updated

**Things to look for during review**

- [ ] Make sure all references to `iot` or `bx` class prefix is using the prefix variable
- [ ] (React) All major areas have a `data-testid` attribute. New test ids should have test written to ensure they are not changed or removed.
- [ ] UI should be checked in RTL mode to ensure the proper handling of layout and text.
- [ ] All strings should be translatable.
- [ ] The code should pass a11y scans (The storybook a11y knob should show no violations). New components should have a11y test written.
- [ ] Unit test should be written and should have a coverage of 90% or higher in all areas.
- [ ] All components should be passing visual regression test. For new styles or components either a visual regression test should be written for all permutations or the base image updated.
- [ ] Changes or new components should either write new or update existing documentation.
- [ ] PR should link and close out an existing issue

### CODE REVIEW

**Review by bjornalm — APPROVED** [review]
Nice work as always. I also tested the deep nesting in https://deploy-preview-3081--carbon-addons-iot-react.netlify.app/?path=/story/1-watson-iot-hierarchylist--with-mixed-hierarchies and it works fine.

Since we are doing this I think it also makes sense to expose an `onExpandedChange` so that consumers can be informed if the "controlled" state is changed by the user in the HierarchyList itself. Could be useful to determine the enabled state of an external "Expand all" button like in the issue example.

**Comment by kevinsperrine** [comment]
> Nice work as always. I also tested the deep nesting in https://deploy-preview-3081--carbon-addons-iot-react.netlify.app/?path=/story/1-watson-iot-hierarchylist--with-mixed-hierarchies and it works fine.
> 
> Since we are doing this I think it also makes sense to expose an `onExpandedChange` so that consumers can be informed if the "controlled" state is changed by the user in the HierarchyList itself. Could be useful to determine the enabled state of an external "Expand all" button like in the issue example.

Good idea. I'll add that.

File: packages/react/src/components/List/HierarchyList/HierarchyList.jsx [hunk 1]
```diff
@@ -276,6 +286,33 @@ const HierarchyList = ({
...
+    }
+    isFirstRender.current = false;
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [expandedIdsProp]);
+
+  /**
+   * effect to trigger calling the onExpandedChange callback. Ignore it on the initial render
+   * when previousExpandedIds is undefined, but fire it on every change afterward
+   */
+  useEffect(() => {
```
Comment (bjornalm) [inline packages/react/src/components/List/HierarchyList/HierarchyList.jsx:305]: Would it not be easier to use the `toggleExpansion` callback from the list, and also keep the APIs more aligned?

Comment (kevinsperrine) [reply packages/react/src/components/List/HierarchyList/HierarchyList.jsx:305]: I started to, but we call setExpandedIds in lots of other places (searches or when the items change, for example), so I thought the effect was cleaner than calling it in multiple places any time it changed.

Comment (bjornalm) [reply packages/react/src/components/List/HierarchyList/HierarchyList.jsx:305]: Ok, makes sense

**Review by davidicus — APPROVED** [review]

**Pull request merged**